### PR TITLE
Add various streaming transducers

### PR DESCRIPTION
**Pull request by satabin**
* [x] Copyless Streaming Tree Transducers which can be used to implement query languages over tree like data
* [x] Streaming Finite State Transducers which can be used for matching inputs and emitting some output
* [x] ~Streaming String Transducers which can be used for matching inputs and emitting some output as soon as possible~

### CODE REVIEW

File: transducers/shared/src/main/scala/fs2/data/stt/Assignment.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+package fs2.data.stt
+
+import cats.Show
+import cats.syntax.show._
+
+sealed trait Assignment[+V <: Variable, +C]
+sealed trait Reset[+C] extends Assignment[Nothing, C]
+object Reset {
+  def unapply[C](r: Reset[C]): Some[Reset[C]] =
```
Comment (ybasket) [inline transducers/shared/src/main/scala/fs2/data/stt/Assignment.scala:25]: Is this really needed? Shouldn't it be a irrefutable pattern anyways and get handled by the compiler?

File: transducers/shared/src/main/scala/fs2/data/stt/STT.scala [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+          env.lookup[F, Expr0[Out]](x).flatMap(loop(_, acc))
+        case Expr0.Char(c) =>
+          F.pure(acc.append(c))
+        case Expr0.Subtree(open, sub, close) =>
+          loop(sub, acc.append(open)).map(_.append(close))
+        case Expr0.Concat(left, right) =>
+          for {
+            acc <- loop(left, acc)
+            acc <- loop(right, acc)
+          } yield acc
```
Comment (ybasket) [inline transducers/shared/src/main/scala/fs2/data/stt/STT.scala:151]:
```suggestion
          loop(left, acc).flatMap(loop(right, _))
```

I find the for-comprehension with re-use of `acc` unnecessarily hard to read and it has the tail `map` problem on Scala 3 IIRC.

File: transducers/shared/src/main/scala/fs2/data/stt/Table.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+object Table {
+
+  implicit object PartialFunctionTable extends Table[PartialFunction] {
+    def get[From, To](m: PartialFunction[From, To])(from: From): Option[To] =
+      m.lift(from)
+  }
+
+  implicit object MapTable extends Table[Map] {
```
Comment (ybasket) [inline transducers/shared/src/main/scala/fs2/data/stt/Table.scala:35]: `Map <: MapOps <: PartialFunction`, so we might run into ambiguous implicit here. Maybe move the partial function instance into a low priority trait to be safe?

**Review by ybasket — COMMENTED** [review]
Not entirely sure what it does without docs, but left some small Scala-specific comments and will stay tuned for updates 😄

File: transducers/shared/src/main/scala/fs2/data/stt/Table.scala (hunk 3 above)
Comment (satabin) [reply transducers/shared/src/main/scala/fs2/data/stt/Table.scala:35]: > Not entirely sure what it does without docs, but left some small Scala-specific comments and will stay tuned for updates smile

Documentation will come. This is a first basic building block for higher-level features. More transducers to come soon, I hope, especially SSTs, which will allow for parsing. This STT can be the base for streaming query languages on tree structures.

File: transducers/shared/src/main/scala/fs2/data/stt/Assignment.scala (hunk 1 above)
Comment (satabin) [reply transducers/shared/src/main/scala/fs2/data/stt/Assignment.scala:25]: How do you plan on writing `case Reset(r) =>` if there is no `unapply` method in object `Reset`? This is basically what the compiler would generate if `Reset` were a case class, IIRC.

Comment (ybasket) [reply transducers/shared/src/main/scala/fs2/data/stt/Assignment.scala:25]: Ah, ignore me. I confused myself while jumping between the trait and the `case class`es implementing it. Hope it works fine on Scala 3 and it changes to `unapply` / introduction of `Matchable`.

**Pull request closed without merging**